src/Compiler/FSCompCheck.fsx

```diff
@@ -0,0 +1,32 @@
+let digits = set { '0' .. '9' }
+
+let rec intStringEndIndex i s =
+    if i >= String.length s then
+        String.length s - 1
+    elif Set.contains s[i] digits then
+        intStringEndIndex (i + 1) s
+    else
+        i - 1
+
+__SOURCE_DIRECTORY__ + "/FSComp.txt"
+|> System.IO.File.ReadAllLines
+|> Seq.fold
+    (fun counters line ->
+        let line = line.TrimStart [| ' '; '#' |]
+        let intStringEndIndex = intStringEndIndex 0 line
+
+        if intStringEndIndex >= 0 then // Line starts with an integer (error code)
+            let errorCode = line[..intStringEndIndex] |> int
+
+            match counters with
+            | [] -> [ errorCode ] // Initialise a counter
+            | head :: tail when errorCode >= head -> errorCode :: tail // Increment current counter
```

Comment:
seems it assumes monotonically increasing error codes, maybe it should construct a set instead?

I don't see a reason for error codes to be in a specific order in the FSComp.txt file, sometimes we may add new errors next to some existing ones, which makes it easier to look at the related error messages from the context of fscomp.txt.

Reply:
But it would make the determination of whether an error code is already used much more non-obvious on review.

Reply:
If we have a list of exception numbers for which the duplicates are allowed, would that solve the issue for those specific messages?

I'm seeing the script as useful to actually generate a new error code (https://github.com/fsharp/fslang-design/blob/main/find.next.id.fsx) and having a unit test that checks the duplicates (excluding those that already exist for historical reasons).

Each of the historical exception would also have a "reason to be" as a string associated in the test.


```fsharp
let errorIdToTolerateDuplicates for = 
  [
     3333, "reason for duplicates existing for error code 3333"
    //...
  ]
  |> dict

// a test that does what this script is doing, but tolerating the duplicates defined above.
```

Reply:
Having this implies we _discourage_ having multiple error messages using the same error code by requiring a reason and treating them as exceptions for historical reasons. 

However, having multiple error messages for the same error can enable better contexts.

e.g. for FS3368 warning against applying a list to an expression without a space in between
```
3368,tcParenThenAdjacentListArgumentReserved,"The syntax '(expr1)[expr2]' is now reserved for indexing and is ambiguous when used as an argument. See https://aka.ms/fsharp-index-notation. If calling a function with multiple curried arguments, add a space between them, e.g. 'someFunction (expr1) [expr2]'."
3368,tcListThenAdjacentListArgumentReserved,"The syntax '[expr1][expr2]' is now reserved for indexing and is ambiguous when used as an argument. See https://aka.ms/fsharp-index-notation. If calling a function with multiple curried arguments, add a space between them, e.g. 'someFunction [expr1] [expr2]'."
3368,tcOtherThenAdjacentListArgumentReserved,"The syntax 'expr1[expr2]' is now reserved for indexing and is ambiguous when used as an argument. See https://aka.ms/fsharp-index-notation. If calling a function with multiple curried arguments, add a space between them, e.g. 'someFunction expr1 [expr2]'."
```

We can just enforce such cases to be next to each other without requiring an explicit reason. We can also avoid accidentally reusing an error code with multiple possible messages, by putting the codes next to each other. If we used exception numbers, we lose this guarantee for those numbers.

Reply:
Do we really need a script for the next id when we can just either scroll to the bottom of this sorted list or fill in an unused code inside this sorted list?
If two PRs use the same error code, individually unique, we can still detect it with this script whereas a find-next-id script won't.

Reply:
@smoothdeveloper That's already the case today. New error codes are already inserted at the bottom for merge conflicts. However, we don't detect conflicting usage of error codes which would lead to one error code being used for logically unrelated errors, leading to mistakes like https://github.com/dotnet/fsharp/pull/12428.

I am the type to look for gaps in existing numbers, but today this path is harder because the codes are unsorted.

Reply:
So I think let's have it like this for now, we can iterate on this script later and it's low risk since it's a CI only thing, at the same time it definitely makes things better than they are now.